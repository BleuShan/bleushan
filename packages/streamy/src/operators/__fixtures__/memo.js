export class EchoIterable {
  [Symbol.iterator]() {
    return this
  }

  next(...value) {
    return !value.length
      ? {
          done: true
        }
      : {
          value,
          done: false
        }
  }
}

export {AsyncIterable} from './of.js'
