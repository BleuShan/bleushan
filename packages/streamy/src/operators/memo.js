import of from './of.js'

export function memo(source) {
  const stream = of(source)

  return stream
}
