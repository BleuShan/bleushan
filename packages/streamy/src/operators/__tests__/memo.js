import {EchoIterable, AsyncIterable} from '../__fixtures__/memo.js'
import {memo} from '../memo.js'

describe.skip('memo', () => {
  let stream
  let source

  beforeEach(() => {
    stream = memo(source)
    jest.spyOn(source, 'next')
  })

  describe('when called on async iterable', () => {
    beforeAll(() => {
      source = new AsyncIterable()
    })

    beforeEach(() => {
      jest.restoreAllMocks()
      source.reset()
    })

    async function iterateAsync() {
      const result = []
      for await (const value of stream) {
        result.push(value)
      }
      return result
    }

    function iterateSync(result) {
      return function iterate() {
        for (const value of stream) {
          result.push(value)
        }
        return result
      }
    }

    it('should not allow synchronous calls if has not completed yet', () => {
      const result = []
      expect(iterateSync(result)).toThrowErrorMatchingSnapshot()
    })

    it('should allow synchronous calls if has completed', async () => {
      const expected = await iterateAsync()
      const result = []

      jest.spyOn(source, 'next')
      expect(stream.completed).toEqual(true)

      expect(iterateSync(result)).not.toThrow()
      expect(source.next).not.toHaveBeenCalled()
      expect(result).toEqual(expected)
    })

    it('should allow asynchronous calls if has completed', async () => {
      const expected = await iterateAsync()

      jest.spyOn(source, 'next')
      expect(stream.completed).toEqual(true)
      const result = await iterateAsync()

      expect(source.next).not.toHaveBeenCalled()
      expect(result).toEqual(expected)
    })
  })

  describe('when called on an iterable that takes an input', () => {
    beforeAll(() => {
      source = new EchoIterable()
    })

    afterEach(() => {
      jest.restoreAllMocks()
    })

    function send(values) {
      const result = []
      for (const value of values) {
        result.push(stream.next(value))
      }

      return result
    }

    describe('when it has memoized the result', () => {
      describe.each([
        [[1, 2, 3]],
        [[{result: 's'}, {test: 2}, undefined]],
        [['string', 'stri', 'hello']],
        [[null, undefined, 0]]
      ])('when it was sent the following: %p', (values) => {
        let expected
        beforeEach(() => {
          expected = send(values)
        })

        it('should not call into the source', () => {
          const result = send(values)
          expect(result).toEqual(expected)
          expect(source.next).toHaveBeenCalledTimes(values.length)
        })
      })
    })
  })
})
