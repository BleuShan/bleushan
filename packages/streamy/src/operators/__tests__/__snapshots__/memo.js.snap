// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`memo when called on async iterable should not allow synchronous calls if has not completed yet 1`] = `"Cannot traverse an asynchronous stream synchronously"`;
